docs/designs/allreduce.md

```diff
@@ -1,14 +1,198 @@
+1. Since the `AllreduceCommunicator` is aware of the failure. It will reconstruct the communicator and re-assign ranks among
+the existing active workers. The existing workers will then continue to run on the tasks at hand.
+1. The master pod will try to create a new worker pod after a specified restart delay period.
+1. Once the worker pod becomes active and `AllreduceCommunicator` is aware of it, we perform the following steps:
+    1. Lock all existing worker pods.
+    1. Send the current model from one of the worker pods to master pod.
+    1. Initialize the model on the new worker pod using the current model that master pod just received.
+    1. All workers continue to operate and perform allreduce operations to update the model.
+
+Note that since the existing active workers have the exact same copy of the model after the previous allreduce operation completes,
+we can guarantee that the new worker will have the same copy of the model as the ones on other workers once the next
+allreduce operation completes.
+
+#### Training with Evaluation
+
+If the worker encounters any evaluation tasks in the above process, it will evaluate the model directly on master once the
+under-going allreduce-based gradients averaging has completed and the model has been updated. The behavior is the same as
+what's described in [model evaluation design doc](model_evaluation.md) except that we are evaluating the model on workers
+instead of on parameter servers. Once an evaluation completes, we send the evaluation result to master for TensorBoard
+service to consume.
+
+#### ElasticDL Embedding Layer
+
+ElasticDL embedding layer is not supported under allreduce-based training since each worker has the exact same copy of
+the model that must fit within each worker's specified resources. Any layers in the model defined via
+``elasticdl.layers.embedding`` will be replaced by `tf.keras.layers.Embedding`.
+
+#### Relevant CLI Arguments
+
+The following CLI arguments are relevant and their behaviors might be changed under allreduce-based training:
+
+* ``--restart_policy``: The pod restart policy when pod crashed.
+the `AllreduceCommunicator` just reconstructed the communicator and we want to wait for a while before restarting the
+failed worker pod which requires reconstruction of the communicator again once the pod becomes active. 
+* ``--distribution_strategy``: In addition to the existing "ParameterServerStrategy" that we have, we add a new strategy
+called "AllreduceStrategy".
+* ``--num_ps_pods`` will be ignored if "AllreduceStrategy" is used and only ``--num_workers`` will be taken into account.
+* ``--use_async`` and ``--lr_staleness_modulation`` will be ignored if "AllreduceStrategy" is used.
+* Only ``--grads_to_wait = 1`` will be supported if "AllreduceStrategy" is used.
+
+The following new CLI arguments are added:
+
+* ``--restart_delay_secs``: The number of seconds to delay before restarting the failed pods. This could be useful when
```

Comment:
Better to use `--relaunch_delay_secs`?

Reply:
I think `restart_delay_secs` might be better to be consistent with existing args like `restart_policy`.